sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java

```diff
@@ -351,25 +351,43 @@ public void readLongs(
       WritableColumnVector c,
       int rowId,
       int level,
-      VectorizedValuesReader data) throws IOException {
+      VectorizedValuesReader data,
+      boolean downCastLongToInt) throws IOException {
     int left = total;
     while (left > 0) {
       if (this.currentCount == 0) this.readNextGroup();
       int n = Math.min(left, this.currentCount);
       switch (mode) {
         case RLE:
           if (currentValue == level) {
-            data.readLongs(n, c, rowId);
+            if (downCastLongToInt) {
+              for (int i = 0; i < n; ++i) {
+                c.putInt(rowId + i, (int) data.readLong());
```

Comment:
I guess I need to understand this better, is this sufficient for both on heap and off heap column vectors?  readLongs function appears to handle both

Reply:
`readLongs` is handling arrays in a way specific to the underlying vector i.e. `OnHeapColumnVector/OffHeapColumnVector` to optimize the operation by doing a memcpy. In our case we can't copy memory chunks because we will have to downcast each value before putting it back.